### ISSUE TITLE
unexpected effect of jw penalty factor

### ISSUE THREAD

**Issue by brooksambrose** [markvanderloo/stringdist#85]
First thank you for an awesome package! The Jaro-Winkler penalty factor isn't behaving as I expect. My understanding is that the JW score is originally a similarity score, and the penalty factor is subtracted to further reduce the similarity between strings that are different in the first four characters.

https://github.com/markvanderloo/stringdist/blob/03cec8adad87d572cd2671789c4fe9e6910dedfa/pkg/src/jaro.c#L114-L124

To make it a distance score you subtract it from 1 in line 118. The penalty is subsequently subtracted on line 123. Does this not reduce the distance (increase similarity) and therefore the penalty is really having the opposite effect of what is intended?

Below I test a five character string with a single character difference. I expect the distance to increase as the penalty factor increases, but here it decreases (i.e. approaches 0, becomes more similar).

I also expect that the penalty should be highest for a difference in the first character, lowest for the fourth, and there should be no additional penalty for the fifth. I'm not sure why the penalty doesn't seem to have any effect if the first character is different `Xxxxx`, it's just the same as `p=0` Jaro distance though I expected it to have a stronger penalty. Also, when the fifth character is different I don't expect there to be an effect at all. 

I may very well be reading the code wrong, or I just am not grocking JW in general, so any advice would be greatly appreciated. Maybe I'm just hung up on the word "penalty" in the stringdist documentation, because it sounds like the higher the penalty the greater the distance. But seeing at least on [Wikipedia](https://en.wikipedia.org/wiki/Jaro–Winkler_distance#Jaro–Winkler_Similarity) as in #52, the p is not called a penalty at all, but rather means "prefix" and actually makes for more "favorable" scores. So in that case, a higher p should indeed decrease distance (increase similarity).

``` r
p<-seq(0,.25,.05);p
#> [1] 0.00 0.05 0.10 0.15 0.20 0.25
sapply(p,function(i) stringdist::stringdist('Xxxxx','xxxxx',method='jw',p=i))
#> [1] 0.1333333 0.1333333 0.1333333 0.1333333 0.1333333 0.1333333
sapply(p,function(i) stringdist::stringdist('xxXxx','xxxxx',method='jw',p=i))
#> [1] 0.13333333 0.12000000 0.10666667 0.09333333 0.08000000 0.06666667
sapply(p,function(i) stringdist::stringdist('xxxxX','xxxxx',method='jw',p=i))
#> [1] 0.13333333 0.10666667 0.08000000 0.05333333 0.02666667 0.00000000
```

<sup>Created on 2020-07-07 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

<details>

<summary>Session info</summary>

``` r
devtools::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value                       
#>  version  R version 3.6.3 (2020-02-29)
#>  os       macOS High Sierra 10.13.6   
#>  system   x86_64, darwin15.6.0        
#>  ui       X11                         
#>  language (EN)                        
#>  collate  en_US.UTF-8                 
#>  ctype    en_US.UTF-8                 
#>  tz       America/Los_Angeles         
#>  date     2020-07-07                  
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package     * version date       lib source        
#>  assertthat    0.2.1   2019-03-21 [1] CRAN (R 3.6.0)
#>  backports     1.1.8   2020-06-17 [1] CRAN (R 3.6.2)
#>  callr         3.4.3   2020-03-28 [1] CRAN (R 3.6.2)
#>  cli           2.0.2   2020-02-28 [1] CRAN (R 3.6.0)
#>  crayon        1.3.4   2017-09-16 [1] CRAN (R 3.6.0)
#>  desc          1.2.0   2018-05-01 [1] CRAN (R 3.6.0)
#>  devtools      2.3.0   2020-04-10 [1] CRAN (R 3.6.2)
#>  digest        0.6.25  2020-02-23 [1] CRAN (R 3.6.0)
#>  ellipsis      0.3.1   2020-05-15 [1] CRAN (R 3.6.2)
#>  evaluate      0.14    2019-05-28 [1] CRAN (R 3.6.0)
#>  fansi         0.4.1   2020-01-08 [1] CRAN (R 3.6.0)
#>  fs            1.4.1   2020-04-04 [1] CRAN (R 3.6.2)
#>  glue          1.4.1   2020-05-13 [1] CRAN (R 3.6.2)
#>  highr         0.8     2019-03-20 [1] CRAN (R 3.6.0)
#>  htmltools     0.5.0   2020-06-16 [1] CRAN (R 3.6.2)
#>  knitr         1.28    2020-02-06 [1] CRAN (R 3.6.0)
#>  magrittr      1.5     2014-11-22 [1] CRAN (R 3.6.0)
#>  memoise       1.1.0   2017-04-21 [1] CRAN (R 3.6.0)
#>  pkgbuild      1.0.8   2020-05-07 [1] CRAN (R 3.6.2)
#>  pkgload       1.1.0   2020-05-29 [1] CRAN (R 3.6.2)
#>  prettyunits   1.1.1   2020-01-24 [1] CRAN (R 3.6.0)
#>  processx      3.4.2   2020-02-09 [1] CRAN (R 3.6.0)
#>  ps            1.3.3   2020-05-08 [1] CRAN (R 3.6.2)
#>  R6            2.4.1   2019-11-12 [1] CRAN (R 3.6.0)
#>  remotes       2.1.1   2020-02-15 [1] CRAN (R 3.6.0)
#>  rlang         0.4.6   2020-05-02 [1] CRAN (R 3.6.2)
#>  rmarkdown     2.3     2020-06-18 [1] CRAN (R 3.6.3)
#>  rprojroot     1.3-2   2018-01-03 [1] CRAN (R 3.6.0)
#>  sessioninfo   1.1.1   2018-11-05 [1] CRAN (R 3.6.0)
#>  stringdist    0.9.5.5 2019-10-21 [1] CRAN (R 3.6.0)
#>  stringi       1.4.6   2020-02-17 [1] CRAN (R 3.6.0)
#>  stringr       1.4.0   2019-02-10 [1] CRAN (R 3.6.0)
#>  testthat      2.3.2   2020-03-02 [1] CRAN (R 3.6.0)
#>  usethis       1.6.1   2020-04-29 [1] CRAN (R 3.6.2)
#>  withr         2.2.0   2020-04-20 [1] CRAN (R 3.6.2)
#>  xfun          0.14    2020-05-20 [1] CRAN (R 3.6.2)
#>  yaml          2.2.1   2020-02-01 [1] CRAN (R 3.6.0)
#> 
#> [1] /Library/Frameworks/R.framework/Versions/3.6/Resources/library
```

</details>

**Comment by markvanderloo** [comment]
Hi Brooks, thank you for your extensive comment. If there's one 'distance' that raises a lot of questions, it is this one :-).

The similarity measure of [winkler (1990)](https://files.eric.ed.gov/fulltext/ED325505.pdf) is given by
```
s_jw = s_j + l * p *(1-s_j)
```
(Winkler sets `p=0.1`) where `l` is the number of equal characters up to maximum four, and `p` is the prefix/penalty factor and `s_j` is the jaro-similarity. All things being equal, a higher `p` means a higher similarity and thus a lower distance as you say. 

So I think that the term 'penalty' factor is not entirely accurate. I will update the documentation.

**Comment by markvanderloo** [comment]
Fixed in 62c3cb88dfa04dc240965afc59ed02a93bd94482
